hs-bindgen/src/HsBindgen/Hs/AST/Name.hs

```diff
@@ -29,50 +62,564 @@ data Namespace =
+      | otherwise     = f c ++ aux isUp cs
+    aux isUp []       = if isUp then "_" else "" -- preserve trailing underscore
+
+-- | Drop invalid characters
+dropInvalidChar :: Char -> String
+dropInvalidChar = const ""
+
+-- | Escape invalid characters
+--
+-- An invalid character transformed to a single quote (@'@) followed by the
+-- Unicode code point (four lowercase hex digits).
+escapeInvalidChar :: Char -> String
+escapeInvalidChar c =
+    let hex = showHex (Char.ord c) ""
+    in  '\'' : replicate (max 0 (4 - length hex)) '0' ++ hex
+
+-- | Valid character predicate
+--
+-- Since 'escapeInvalidChar' uses the single quote (@'@) as an escape
+-- character, it is treated specially.  It should never occur in a 'CName',
+-- and this predicate does not consider it valid.
+isValidChar :: Char -> Bool
+isValidChar c = Char.isAlphaNum c || c == '_'
+
+-- | Join parts of a name by concatenating them
+joinWithConcat :: [Text] -> Text
+joinWithConcat = T.concat
+
+-- | Join parts of a name with underscores (@_@)
+joinWithSnakeCase :: [Text] -> Text
+joinWithSnakeCase = T.intercalate "_"
+
+-- | Join parts of a name in @camelCase@ style
+--
+-- The first character of all parts but the first is changed to uppercase (if it
+-- is a letter), and the results are concatenated.
+--
+-- Since this function may change the case of letters, it can cause name
+-- collisions when different C names only differ by case of the first letter.
+joinWithCamelCase :: [Text] -> Text
+joinWithCamelCase = \case
+    (t:ts) -> T.concat $ t : map upperFirstChar ts
+    []     -> T.empty
+  where
+    upperFirstChar :: Text -> Text
+    upperFirstChar t = case T.uncons t of
+      Just (c, t') -> T.cons (Char.toUpper c) t'
+      Nothing      -> t
+
+-- | Do not handle reserved names
+handleReservedNone :: HsName ns -> HsName ns
+handleReservedNone = id
+
+-- | If a name is in a list of reserved names, transform it
+--
+-- The transformation function must return a valid Haskell name.  One
+-- transformation function is provided in this module: 'appendSingleQuote'.
+--
+-- For example, 'keywords' may be used to avoid using Haskell keywords as
+-- identifiers.
+handleReservedNames :: (Text -> Text) -> [Text] -> HsName ns -> HsName ns
+handleReservedNames f reserved name@(HsName t)
+    | t `elem` reserved = HsName $ f t
+    | otherwise         = name
+
+-- | Append a single quote (@'@) to a name
+appendSingleQuote :: Text -> Text
+appendSingleQuote = (<> "'")
+
+-- | Haskell keywords
+--
+-- <https://wiki.haskell.org/Keywords>
+keywords :: [Text]
+keywords =
+    [ "as"
+    , "case"
+    , "class"
+    , "data"
+    , "default"
+    , "deriving"
+    , "do"
+    , "else"
+    , "family"
+    , "forall"
+    , "foreign"
+    , "hiding"
+    , "if"
+    , "import"
+    , "in"
+    , "infix"
+    , "infixl"
+    , "infixr"
+    , "instance"
+    , "let"
+    , "mdo"
+    , "module"
+    , "newtype"
+    , "of"
+    , "proc"
```

Comment:
also `role`

Reply:
and `pattern` and `static`

Reply:
Thank you!  I (rebased and) added these.

> This commit replaces `keywords` with separate `varReservedNames` and `typeVarReservedNames`, for variables and type variables, respectively.
>
> Reserved names `pattern` and `static` are added to `varReservedNames`.
>
> List `typeVarReservedNames` contains reserved name `role` as well as all `varReservedNames`.  `role` is only problematic when used as a type variable.
>
> Documentation is added to reference these added reserved names.